Automattic/super-cool-ad-inserter-plugin
# Add new "start" setting to define number of paragraphs before first insertion

Adds a new option in plugin settings which lets editors set the number of paragraphs before the first insertion, independently of the number of paragraphs between insertions.

Closes #53.

To test:

1. Before checking out this branch, ensure that you have [created](https://github.com/Automattic/super-cool-ad-inserter-plugin/blob/master/docs/adding-ads.md#manual-ad-creation) at least several (>=4) Ad Position widgets.
2. Go to **Settings > Ad Inserter** and note the setting of "Number of paragraphs between insertions" (`3` by default).
3. View a post with enough content to render all widgets, and note positions.
4. Now check out this branch.

### Control test
1. In **Settings > Ad Inserter**, observe a new "Number of paragraphs before first insertion" setting that is set to `3` by default. 
2. Refresh the post and observe that the widget positions are unchanged from before checking out the branch.

### Test functionality of new setting
1. In **Settings > Ad Inserter**, change the "Number of paragraphs before first insertion" to `2` (something lower than `3`).
2. Refresh the post and observe that the first widget gets inserted after the second paragraph, and the subsequent widgets get inserted every 3 paragraphs (or whatever "Number of paragraphs between insertions" is set to). 
3. In **Settings > Ad Inserter**, change the "Number of paragraphs before first insertion" to `6` (something higher than `3`).
4. Refresh the post and observe that the first widget gets inserted after the sixth paragraph, and the subsequent widgets get inserted every 3 paragraphs (or whatever "Number of paragraphs between insertions" is set to).
5. Change the "Number of paragraphs between insertions" setting to something else, refresh the post, and verify that the first widget still gets inserted after the sixth paragraph, but the subsequent widgets get inserted after the number of paragraphs of the new "Number of paragraphs between insertions" setting.

### Test functionality with not enough content
1. Create a new post or edit the post content to shorten the number of paragraphs to 5, or something less than the "Number of paragraphs before first insertion" setting.
2. View the short post and observe that no ads get inserted.

### Test interaction with other settings
1. In **Settings > Ad Inserter**, change the "Minimum number of paragraphs needed in a post to insert ads" setting to `6` (something higher than the number of paragraphs in your short post).
2. Change the "Number of paragraphs before first insertion" to `2`(something lower than the number of paragraphs in your short post).
3. View the short post and observe that no ads get inserted.
4. Increase the "Minimum number of paragraphs needed in a post to insert ads" setting to `3` (something lower than the number of paragraphs in your short post + 1).
5. Refresh the short post and observe that a widget now gets inserted after the second paragraph.

### Test that at least one paragraph must be rendered after the last inserted widget
1. Set "Minimum number of paragraphs needed in a post to insert ads" to `5` (equal to the number of paragraphs in your short post).
2. View the short post and observe that no ads get inserted.
3. Set "Minimum number of paragraphs needed in a post to insert ads" to `2` (lower than the number of paragraphs in your short post) and "Minimum number of paragraphs needed in a post to insert ads" to `5` (equal to the number of paragraphs in your short post).
4. View the short post and observe that no ads get inserted.
